src/runtime/testing.ts

```diff
@@ -42,46 +42,69 @@ declare global {
 
 export type TestContext = NexusTestContextRoot
 
+export interface CreateTestContextOptions {
+  /**
+   * A path to the entrypoint of your app. Only necessary if the entrypoint falls outside of Nexus conventions.
+   * You should typically use this if you're using `nexus dev --entrypoint` or `nexus build --entrypoint`.
+   */
+  entrypointPath?: string
+}
+
 /**
  * Setup a test context providing utilities to query against your GraphQL API
  *
  * @example
  *
  * With jest
  * ```
- * import { setupTest, TestContext } from 'nexus/testing'
+ * import { createTestContext, TestContext } from 'nexus/testing'
  *
- * let testCtx: TestContext
+ * let ctx: TestContext
  *
  * beforeAll(async () => {
- *  testCtx = await setupTest()
- *  await testCtx.server.start()
+ *  ctx = await createTestContext()
+ *  await ctx.server.start()
  * })
  *
  * afterAll(async () => {
- *  await testCtx.server.stop()
+ *  await ctx.server.stop()
  * })
  * ```
  */
-
-export async function createTestContext(): Promise<TestContext> {
+export async function createTestContext(opts?: CreateTestContextOptions): Promise<TestContext> {
   // Guarantee that development mode features are on
   process.env.NEXUS_STAGE = 'dev'
 
   // todo figure out some caching system here, e.g. imagine jest --watch mode
-  const layout = await Layout.create()
+  const layout = await Layout.create({ entrypointPath: opts?.entrypointPath })
   const pluginManifests = await Plugin.getUsedPlugins(layout)
   const randomPort = await getPort({ port: getPort.makeRange(4000, 6000) })
-  const app = require('../index') as PrivateApp
+  const app = require('../index').default as PrivateApp
+
+  const forcedServerSettings = {
+    port: randomPort,
+    playground: false, // Disable playground during tests
+    startMessage() {}, // Make server silent
+  }
+  const originalSettingsChange = app.settings.change
 
   app.settings.change({
-    server: {
-      port: randomPort,
-      playground: false, // Disable playground during tests
-      startMessage() {}, // Make server silent
-    },
+    server: forcedServerSettings,
   })
 
+  /**
+   * If app ever calls app.settings.change, force some server settings anyway
+   */
+  app.settings.change = (newSettings) => {
+    if (newSettings.server !== undefined) {
+      newSettings.server = {
+        ...newSettings.server,
+        ...forcedServerSettings,
+      }
+    }
+    originalSettingsChange(newSettings)
+  }
```

Comment:
This got me thinking... e.g. a) ugh what a mess blah blah b) there must be a better way c) thanks for dealing with this! d) yikes architecture.

a b and d are directed at me!

Nice catch

Reply:
a), b) and d) don't have to be directed at you only, we came up with that architecture together 🙈.

Right now I see this as a pure side-effect of having configuration as an API, maybe I'm wrong, but I don't see a better way to handle that (as hacky as it is)